### PR TITLE
Add POST user endpoint

### PR DESCRIPTION
**Pull request by danielchen115**
### Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/POST-user-endpoint-17370b16be7c4b7d918840a70ccd75d2)

### Implementation description

-

### Steps to test

1. Make sure the user does not exist in firebase or local db
2. hit the `create` endpoint with a email in the body
3. check if the user exists in Firebase and in our local db with matching firebase UIDs

### What should reviewers focus on?

- Any edge cases where firebase/local failures can cause an unstable state for user management

### CODE REVIEW

**Comment by Tayef-Shah** [comment]
What's the command to get an auth token again?

File: accounts/tests/views/test_user.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        response = self.client.put(url)
+        self.assertEqual(response.status_code, status.HTTP_405_METHOD_NOT_ALLOWED)
+
+        response = self.client.patch(url)
+        self.assertEqual(response.status_code, status.HTTP_405_METHOD_NOT_ALLOWED)
+
+        response = self.client.delete(url)
+        self.assertEqual(response.status_code, status.HTTP_405_METHOD_NOT_ALLOWED)
+
+    def test_unauthorized(self):
```
Comment (Tayef-Shah) [inline accounts/tests/views/test_user.py:40]: get request should also return 401

File: accounts/tests/serializers/test_user_create.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        serializer = UserCreateSerializer(data=self.data)
+        self.assertTrue(serializer.is_valid())
+        self.assertRaises(ValidationError, serializer.save)
+        mock_create.assert_called_once()
+
+    @patch("firebase_admin.auth.create_user")
+    @patch("firebase_admin.auth.delete_user")
+    @patch("accounts.models.User.objects.create", side_effect=Exception(""))
+    def test_user_create__local_user_exists(
+        self, mock_create, mock_delete, mock_model_create
```
Comment (brittanylau) [inline accounts/tests/serializers/test_user_create.py:44]: python applies decorators bottom up, so the order should be reversed here (rn `mock_model_create` references `firebase_admin.auth.create_user`)

File: accounts/tests/serializers/test_user_create.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from firebase_admin import auth
+from rest_framework.serializers import ValidationError
+
+
+class UserCreateSerializerTestCase(TestCase):
+    def setUp(self):
+        self.data = {"email": "test@asd.com"}
+
+    @patch("firebase_admin.auth.create_user")
+    def test_user_create(self, mock_create):
```
Comment (brittanylau) [inline accounts/tests/serializers/test_user_create.py]: since we're overriding the behaviour of the create method, can we verify that we still create the django user object?

File: accounts/tests/serializers/test_user_create.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        self.assertTrue(serializer.is_valid())
+        user = serializer.save()
+        self.assertEqual(user.firebase_uid, uid)
+        mock_create.assert_called_once()
+
+    @patch(
+        "firebase_admin.auth.create_user",
+        side_effect=auth.EmailAlreadyExistsError("message", "cause", "response"),
+    )
+    def test_user_create__firebase_user_exists(self, mock_create):
```
Comment (brittanylau) [inline accounts/tests/serializers/test_user_create.py]: can we also verify that we don't create a django user obj when firebase raises an exception?

File: accounts/tests/serializers/test_user_create.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    def test_user_create(self, mock_create):
+        uid = "test_uid"
+        serializer = UserCreateSerializer(data=self.data)
+        # anonymous object creation
+        firebase_user = type("", (object,), {"uid": uid})()
+        mock_create.return_value = firebase_user
+        self.assertTrue(serializer.is_valid())
+        user = serializer.save()
+        self.assertEqual(user.firebase_uid, uid)
+        mock_create.assert_called_once()
```
Comment (brittanylau) [inline accounts/tests/serializers/test_user_create.py]: nit: can we be extra granular and use `assert_called_once_with(args)` instead of `assert_called_once()`? (just to be extra careful with auth stuff)

File: accounts/tests/views/test_user.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        url = reverse("users-list")
+        self.client.force_authenticate(self.user)
+        response = self.client.post(
+            url,
+            {
+                "email": "test@user.com",
+            },
+            format="json",
+        )
+        self.assertTrue(mock_method.called)
```
Comment (brittanylau) [inline accounts/tests/views/test_user.py]: nit: can we do `assert_called_once_with(args)`?

File: accounts/tests/views/test_user.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+    def test_method_not_allowed(self):
+        url = reverse("users-list")
```
Comment (brittanylau) [inline accounts/tests/views/test_user.py]: put, patch, and delete are applied to `users-detail` ([docs](https://www.django-rest-framework.org/api-guide/routers/#defaultrouter)), so these would continue to pass even if we added them! this used to be inconsistent in our tests but fixed in #57 and #58

File: accounts/urls.py [hunk 8]
```diff
@@ -3,7 +3,7 @@
 from .views import UserViewSet
 
 router = DefaultRouter()
-router.register(r"users", UserViewSet)
+router.register(r"users", UserViewSet, basename="users")
```
Comment (brittanylau) [inline accounts/urls.py]: nit: we removed basenames in #57 since we're not rly using them – would be nice to be consistent! sorry for not communicating this earlier 😓

**Review by brittanylau — APPROVED** [review]
logic looks very noice! a few suggestions to make tests tighter

**Comment by brittanylau** [comment]
thanks @danielchen115!! I think it's failing the CI build due to that incompatibility with the faker library on the latest python versions 😖 we specified the compatible version in #64 so might need to rebase
